### ISSUE TITLE
Random pipeline operations replicated identically over multiple GPU batches

### ISSUE THREAD

**Issue by paulbisso** [NVIDIA/DALI#426]
I'm using ops.CoinFlip() as a mask for my random rotate and flip operations in a training pipeline (code below). The expected behavior (i.e. what I would expect to happen) is that the tensor of 1s and 0s produced  is generated anew for _every individual GPU-batch_. However, what's happening is that a single tensor is being generated and applied as an identical mask for multiple GPU-batches. And unfortunately, it's not just the mask -- it's the combination of mask+transformation details. For example, if I'm using 8 GPUs and a batch size of 4 images (per GPU), the first 5-6 sets of 4 images, representing the first 5-6 GPUs, will have applied identical sequences of transformations. In other words, for each set of 4 images, the 2nd image in each set, say, would have a 34 degree rotation applied.

This is problematic because it introduces an extraneous pattern into my dataset (i.e. an identical transformation which is applied every so many images). I think this is why when I train using the PyTorch dataloader vs. the DALI dataloader, with all else held identical, I'm seeing better training using the PyTorch dataloader.

I'm running this on in a jupyter notebook on an AWS p2.8xlarge, AMI -- Deep Learning AMI (Ubuntu) Version 20.0 (ami-0c9ae74667b049f59). 

My env.yml file is here (zipped): [env.zip](https://github.com/NVIDIA/DALI/files/2749992/env.zip)

And code to reproduce with screen grabs is below: 

Training pipeline:
```python
class TrainPipeline(Pipeline):
        def __init__(self, batch_size, num_threads, device_id, image_dir):
            super(TrainPipeline, self).__init__(batch_size, num_threads, device_id)
            self.input = ops.ExternalSource()
            self.input_label = ops.ExternalSource()
            self.decode = ops.nvJPEGDecoder(device = "mixed", output_type = types.RGB)
            self.resize = ops.Resize(device="gpu")
            self.resize_rng = ops.Uniform(range = (256,512))
            self.crop = ops.Crop(device="gpu", crop=(224,224))
            self.rotate = ops.Rotate(device="gpu") # Need to specify angle
            self.angle_rng = ops.Uniform(range = (-45.0,45.0))
            self.coin = ops.CoinFlip(probability=0.5)
            self.coin2 = ops.CoinFlip(probability=0.5)
            self.coin3 = ops.CoinFlip(probability=0.5)
            self.vertflip = ops.Flip(device="gpu", vertical=1)
            self.horizflip = ops.Flip(device="gpu")
            self.norm = ops.NormalizePermute(device = "gpu",
                                            height = 224,
                                            width = 224,
                                            mean = [123.675, 116.28, 103.53],
                                            std = [58.395, 57.12, 57.375])

        def define_graph(self):
            self.jpegs = self.input()
            self.labels = self.input_label()
            images = self.decode(self.jpegs)
            images = self.resize(images, resize_shorter=self.resize_rng())
            images = self.crop(images)
            images = self.vertflip(images, mask=self.coin())
            images = self.rotate(images, angle = self.angle_rng(), mask = self.coin2())
            images = self.horizflip(images, mask=self.coin3())
            images = self.norm(images)
            return (images, self.labels)

        def iter_setup(self):
            (images, labels) = iter_train.next()
            self.feed_input(self.jpegs, images)
            self.feed_input(self.labels, labels)
```
Iterator and pipeline setup:
```python
def setup_dali_iterators(train_batch):
    # Loaders
    train_dir = 'data/train'
    # Build the iterators. 
    eii_train = ExternalInputIterator(batch_size=train_batch, image_dir=train_dir, shuffle_flag=True)
    iter_train = iter(eii_train)
    
    return iter_train
    
def setup_data_loader(num_devices, use_dali, train_batch iter_train):
    if num_devices > 0 and use_dali:
        # GPUs
        N = num_devices
        # Loaders
        train_dir = 'data/train'

        # Training loader
        pipe_t =  [TrainPipeline, (0, 2)]

        pipe_name, label_range = pipe_t
        print ("RUN: "  + pipe_name.__name__)
        pipes = [pipe_name(batch_size=train_batch, num_threads=2, device_id = device_id, image_dir = train_dir) for device_id in range(N)]

        train_loader = DALIGenericIterator(pipes, ['data', 'label'], num_train_images, stop_at_epoch = False)
        return {'train': train_loader}
```

Plotting function:
```python
import matplotlib.pyplot as _plt
import numpy as _np
import matplotlib.gridspec as _gridspec

def view_DALIiter_images(batch_data, batch_size, normalized):
    # batch_data - the direct image output of running .next() on a DALI pipeline
    # batch_size - the number of images in each batch per GPU
    # normalized - Boolean flag denoting whether the image has undergone normalization (of the kind used for all pretrained Pytorch models)

    num_gpus = len(batch_data)
    columns = 4
    rows = (batch_size + 1) // (columns)

    gs = _gridspec.GridSpec(rows, columns)
    # Account for the fact that the image coming in may be sized as HWC vs CHW.
    if batch_data[0]['data'][0].size()[0] == 3:
        CHW_flag = 1
    else:
        CHW_flag = 0
    for i in range(num_gpus):
        titlestr = "GPU # {}".format(i)
        fig = _plt.figure(figsize = (32,(32 // columns) * rows))
        _plt.title(titlestr)
        for j in range(batch_size):
            _plt.subplot(gs[j])
            _plt.axis("off")
            # If the image is coming in as CHW, it has to be reshaped to HWC
            thisimg = batch_data[i]['data'][j].cpu().numpy()
            if CHW_flag:
                thisimg = _np.moveaxis(thisimg,0,-1)
            # If the image has been normalized (assuming Alexnet-type normalization on uint8s ranging from 0 to 255 originally),
            # then de-normalize it.
            if normalized:
                thisimg[:,:,0] = thisimg[:,:,0]*58.395+123.675
                thisimg[:,:,1] = thisimg[:,:,1]*57.12+116.28
                thisimg[:,:,2] = thisimg[:,:,2]*57.375+103.53
                thisimg = _np.uint8(thisimg)
            # Plot.
            _plt.imshow(thisimg)
```

Build pipelines, grab a batch (of 32 -- 4 x 8 GPUs) and plot.
```python
if use_dali:
    iter_train = setup_dali_iterators(4)
else:
    iter_train = []
    iter_valid = []
    iter_test = []
loader_dict = setup_data_loader(num_devices, use_dali, 4, iter_train)
imgs = loader_dict['train'].next()
view_DALIiter_images(imgs,4,True)
```
And here's what results -- 4 GPUs with identical rotation angles and masks (each row represents a batch of 4 images from 1 GPU). The other 4 GPUs had different rotation angles and masks, but were identical among themselves. Any thoughts on a workaround? 

![dali_ex_rotate](https://user-images.githubusercontent.com/41169757/51041459-fbc0de00-1587-11e9-9ccb-2c29d0247b2e.png)

**Comment by JanuszL** [comment]
Hi,
This seems strange. There is `seed` argument you can provide to the pipeline constructor. You are not doing that so the default value `-1` should be passed `time(0)` should be used as a seed. Maybe all calls returns the same value - as construction of all pipelines takes less than 1s together and they obtain the same seed.
Please try to provide manually different seed to each instance of the pipeline and see how that works for you.

**Comment by paulbisso** [comment]
Thanks @JanuszL that seemed to do the trick. The 1 s interval definitely isn't fast enough to be used as a random seed (esp. as a default setting), since most pipelines will be constructed near-simultaneously when using Python's list comprehension.

Below is the code that got it working. Note that I had to multiply Python's time.time() by 10,000,000 before converting to int() in order to ensure that I don't get duplicate values.

```python
class TrainPipeline(Pipeline):
        def __init__(self, batch_size, num_threads, device_id, image_dir, seed):
            super(TrainPipeline, self).__init__(batch_size, num_threads, device_id, seed=seed)
            self.input = ops.ExternalSource()
            self.input_label = ops.ExternalSource()
            self.decode = ops.nvJPEGDecoder(device = "mixed", output_type = types.RGB)
            self.resize = ops.Resize(device="gpu")
            self.resize_rng = ops.Uniform(range = (256,512))
            self.crop = ops.Crop(device="gpu", crop=(224,224))
            self.rotate = ops.Rotate(device="gpu") # Need to specify angle
            self.angle_rng = ops.Uniform(range = (-45.0,45.0))
            self.coin = ops.CoinFlip(probability=0.5)
            self.coin2 = ops.CoinFlip(probability=0.5)
            self.coin3 = ops.CoinFlip(probability=0.5)
            self.vertflip = ops.Flip(device="gpu", vertical=1)
            self.horizflip = ops.Flip(device="gpu")
            self.norm = ops.NormalizePermute(device = "gpu",
                                            height = 224,
                                            width = 224,
                                            mean = [123.675, 116.28, 103.53],
                                            std = [58.395, 57.12, 57.375])

        def define_graph(self):
            self.jpegs = self.input()
            self.labels = self.input_label()
            images = self.decode(self.jpegs)
            images = self.resize(images, resize_shorter=self.resize_rng())
            images = self.crop(images)
            images = self.vertflip(images, mask=self.coin())
            images = self.rotate(images, angle = self.angle_rng(), mask = self.coin2())
            images = self.horizflip(images, mask=self.coin3())
            images = self.norm(images)
            return (images, self.labels)

        def iter_setup(self):
            (images, labels) = iter_train.next()
            self.feed_input(self.jpegs, images)
            self.feed_input(self.labels, labels)
```

```python
def setup_dali_iterators(train_batch):
    # Loaders
    train_dir = 'data/train'
    # Build the iterators. 
    eii_train = ExternalInputIterator(batch_size=train_batch, image_dir=train_dir, shuffle_flag=True)
    iter_train = iter(eii_train)
    
    return iter_train
    
def setup_data_loader(num_devices, use_dali, train_batch iter_train):
    if num_devices > 0 and use_dali:
        # GPUs
        N = num_devices
        # Loaders
        train_dir = 'data/train'

        # Training loader
        pipe_t =  [TrainPipeline, (0, 2)]

        pipe_name, label_range = pipe_t
        print ("RUN: "  + pipe_name.__name__)
        pipes = [pipe_name(batch_size=train_batch, num_threads=2, device_id = device_id, image_dir = train_dir, seed = int(time.time()*10000000)) for device_id in range(N)]

        train_loader = DALIGenericIterator(pipes, ['data', 'label'], num_train_images, stop_at_epoch = False)
        return {'train': train_loader}
```

**Comment by JanuszL** [comment]
It should fix the problem in the future https://github.com/NVIDIA/DALI/pull/430

**Comment by JanuszL** [comment]
Done
